File: ocaml/test_book.ml
(** Tests for the OCaml matching engine.

    These mirror the invariants the Python engine is tested against: price
    priority, FIFO time priority within a level, execution at the maker's price
    (price improvement when walking the book), correct resting of remainders,
    and that market orders never rest. [Book.check_invariants] runs after every
    mutation, so a structural violation fails the test that caused it. *)

open Orders

let failures = ref 0
let checks = ref 0

let check name cond =
  incr checks;
  if not cond then begin
    incr failures;
    Printf.printf "  FAIL  %s\n" name
  end

let check_int name ~expected ~actual =
  incr checks;
  if expected <> actual then begin
    incr failures;
    Printf.printf "  FAIL  %s (expected %d, got %d)\n" name expected actual
  end

let fills events =
  List.filter_map (function Fill f -> Some f | _ -> None) events

let accepteds events =
  List.filter_map (function Accepted a -> Some a | _ -> None) events

let cancels events =
  List.filter_map (function Cancelled c -> Some c | _ -> None) events

(* --------------------------------------------------------------- fixtures *)

(** A book with asks resting at 101, 102 and bids at 99, 98. *)
let two_sided () =
  let b = Book.create () in
  ignore (Book.submit_limit b ~owner:"mm" ~side:Sell ~price:101 ~qty:10 ~time:0.);
  ignore (Book.submit_limit b ~owner:"mm" ~side:Sell ~price:102 ~qty:10 ~time:0.);
  ignore (Book.submit_limit b ~owner:"mm" ~side:Buy ~price:99 ~qty:10 ~time:0.);
  ignore (Book.submit_limit b ~owner:"mm" ~side:Buy ~price:98 ~qty:10 ~time:0.);
  Book.check_invariants b;
  b

(* ------------------------------------------------------------------ tests *)

let test_best_prices_and_mid () =
  let b = two_sided () in
  check "best bid is 99" (Book.best_bid b = Some 99);
  check "best ask is 101" (Book.best_ask b = Some 101);
  check "mid is 100" (Book.mid b = Some 100.);
  check "spread is 2" (Book.spread b = Some 2)

let test_price_priority () =
  (* A buy for 10 must hit the 101 ask, not the 102 one. *)
  let b = two_sided () in
  let _, events = Book.submit_market b ~owner:"taker" ~side:Buy ~qty:10 ~time:1. in
  Book.check_invariants b;
  let fs = fills events in
  check_int "one fill" ~expected:1 ~actual:(List.length fs);
  check_int "filled at the best ask" ~expected:101 ~actual:(List.hd fs).f_price;
  check "101 level is gone" (Book.best_ask b = Some 102)

let test_time_priority_is_fifo () =
  (* Two makers at the same price: the earlier one must fill first. *)
  let b = Book.create () in
  let first, _ = Book.submit_limit b ~owner:"early" ~side:Sell ~price:101 ~qty:5 ~time:0. in
  let second, _ = Book.submit_limit b ~owner:"late" ~side:Sell ~price:101 ~qty:5 ~time:0. in
  let _, events = Book.submit_market b ~owner:"taker" ~side:Buy ~qty:5 ~time:1. in
  Book.check_invariants b;
  let fs = fills events in
  check_int "one fill" ~expected:1 ~actual:(List.length fs);
  check_int "earlier order filled" ~expected:first ~actual:(List.hd fs).f_maker_order_id;
  check "later order still resting" (Book.get_order b second <> None)

let test_walks_the_book_with_price_improvement () =
  (* A large marketable buy consumes 101 then 102, printing each at the
     maker's price rather than at a single blended price. *)
  let b = two_sided () in
  let _, events = Book.submit_limit b ~owner:"taker" ~side:Buy ~price:102 ~qty:15 ~time:1. in
  Book.check_invariants b;
  let fs = fills events in
  check_int "two fills" ~expected:2 ~actual:(List.length fs);
  check_int "first prints at 101" ~expected:101 ~actual:(List.nth fs 0).f_price;
  check_int "first fills 10" ~expected:10 ~actual:(List.nth fs 0).f_qty;
  check_int "second prints at 102" ~expected:102 ~actual:(List.nth fs 1).f_price;
  check_int "second fills 5" ~expected:5 ~actual:(List.nth fs 1).f_qty;
  check "nothing rested: order fully filled" (accepteds events = [])

let test_marketable_limit_rests_remainder () =
  (* Buy 25 at 101 with only 10 available there: 10 fill, 15 rest as the bid. *)
  let b = two_sided () in
  let _, events = Book.submit_limit b ~owner:"taker" ~side:Buy ~price:101 ~qty:25 ~time:1. in
  Book.check_invariants b;
  check_int "one fill" ~expected:1 ~actual:(List.length (fills events));
  let a = List.hd (accepteds events) in
  check_int "remainder rested" ~expected:15 ~actual:a.a_qty;
  check "remainder is the new best bid" (Book.best_bid b = Some 101)

let test_market_order_never_rests () =
  (* Demand 50 against 20 of resting asks: 20 fill, 30 are discarded. *)
  let b = two_sided () in
  let _, events = Book.submit_market b ~owner:"taker" ~side:Buy ~qty:50 ~time:1. in
  Book.check_invariants b;
  check_int "two fills" ~expected:2 ~actual:(List.length (fills events));
  let c = List.hd (cancels events) in
  check_int "remainder discarded" ~expected:30 ~actual:c.c_qty;
  check "reported as unfilled_market" (c.c_reason = "unfilled_market");
  check "ask side is empty" (Book.best_ask b = None);
  check "taker did not rest a bid" (Book.best_bid b = Some 99)

let test_partial_fill_keeps_priority () =
  (* A maker partially filled stays at the head of its level with the
     remaining quantity. *)
  let b = Book.create () in
  let maker, _ = Book.submit_limit b ~owner:"mm" ~side:Sell ~price:101 ~qty:10 ~time:0. in
  ignore (Book.submit_market b ~owner:"taker" ~side:Buy ~qty:4 ~time:1.);
  Book.check_invariants b;
  match Book.get_order b maker with
  | None -> check "maker still resting" false
  | Some o -> check_int "remaining quantity" ~expected:6 ~actual:o.qty

let test_cancel () =
  let b = two_sided () in
  let id, _ = Book.submit_limit b ~owner:"mm" ~side:Buy ~price:100 ~qty:7 ~time:0. in
  check "new order is the best bid" (Book.best_bid b = Some 100);
  (match Book.cancel b ~order_id:id ~time:1. with
  | None -> check "cancel returned an event" false
  | Some c ->
      check_int "cancelled quantity" ~expected:7 ~actual:c.c_qty;
      check "reason is user" (c.c_reason = "user"));
  Book.check_invariants b;
  check "best bid reverts to 99" (Book.best_bid b = Some 99);
  check "second cancel is a no-op" (Book.cancel b ~order_id:id ~time:2. = None)

let test_non_crossing_limit_just_rests () =
  let b = two_sided () in
  let _, events = Book.submit_limit b ~owner:"mm" ~side:Buy ~price:100 ~qty:5 ~time:1. in
  Book.check_invariants b;
  check "no fills" (fills events = []);
  check_int "one acceptance" ~expected:1 ~actual:(List.length (accepteds events));
  check "book is not crossed" (Book.best_bid b = Some 100 && Book.best_ask b = Some 101)

let test_depth_is_aggregated_best_first () =
  let b = two_sided () in
  ignore (Book.submit_limit b ~owner:"mm2" ~side:Sell ~price:101 ~qty:3 ~time:0.);
  Book.check_invariants b;
  match Book.depth b Sell None with
  | (p0, q0) :: (p1, _) :: _ ->
      check_int "best ask level first" ~expected:101 ~actual:p0;
      check_int "quantities aggregated" ~expected:13 ~actual:q0;
      check_int "next level after" ~expected:102 ~actual:p1
  | _ -> check "two ask levels present" false

let test_pre_trade_mid_is_recorded_before_the_trade () =
  (* The mid recorded on a fill must be the one prevailing before the incoming
     order moved the book, which is what spread-capture accounting needs. *)
  let b = two_sided () in
  let _, events = Book.submit_market b ~owner:"taker" ~side:Buy ~qty:10 ~time:1. in
  let f = List.hd (fills events) in
  check "pre-trade mid is 100" (f.f_pre_trade_mid = Some 100.)

let test_rejects_non_positive_quantity () =
  let b = Book.create () in
  let raised =
    try
      ignore (Book.submit_limit b ~owner:"mm" ~side:Buy ~price:100 ~qty:0 ~time:0.);
      false
    with Invalid_argument _ -> true
  in
  check "zero quantity is rejected" raised

(* ------------------------------------------------------------------- main *)

let () =
  let tests =
    [
      ("best prices and mid", test_best_prices_and_mid);
      ("price priority", test_price_priority);
      ("time priority is FIFO", test_time_priority_is_fifo);
      ("walks the book with price improvement", test_walks_the_book_with_price_improvement);
      ("marketable limit rests remainder", test_marketable_limit_rests_remainder);
      ("market order never rests", test_market_order_never_rests);
      ("partial fill keeps priority", test_partial_fill_keeps_priority);
      ("cancel", test_cancel);
      ("non-crossing limit just rests", test_non_crossing_limit_just_rests);
      ("depth is aggregated, best first", test_depth_is_aggregated_best_first);
      ("pre-trade mid recorded before trade", test_pre_trade_mid_is_recorded_before_the_trade);
      ("rejects non-positive quantity", test_rejects_non_positive_quantity);
    ]
  in
  List.iter
    (fun (name, f) ->
      Printf.printf "%-42s" name;
      let before = !failures in
      f ();
      if !failures = before then print_string "ok\n" else print_newline ())
    tests;
  Printf.printf "\n%d checks, %d failures\n" !checks !failures;
  if !failures > 0 then exit 1

File: tests/test_book.py
"""Scenario tests for the matching engine.

Every economically meaningful rule gets an explicit, hand-checkable scenario:
price priority, FIFO time priority, partial fills, walking the book, price
improvement, marketable limits, market-order remainders, and cancel semantics.
The differential/property tests in test_book_properties.py then cover the
combinatorial space; these tests are the human-readable specification.
"""

import pytest

from market_maker_sim.book import LimitOrderBook
from market_maker_sim.orders import Accepted, Cancelled, Fill, Side

B, S = Side.BUY, Side.SELL


def fills(events):
    return [e for e in events if isinstance(e, Fill)]


def make_book():
    return LimitOrderBook()


class TestResting:
    def test_limit_order_rests_and_is_accepted(self):
        book = make_book()
        oid, events = book.submit_limit("a", B, 100, 5, time=1.0)
        assert events == [Accepted(time=1.0, order_id=oid, owner="a", side=B,
                                   price=100, qty=5)]
        assert book.best_bid() == 100
        assert book.best_ask() is None
        assert book.get_order(oid).qty == 5
        book.check_invariants()

    def test_mid_and_spread(self):
        book = make_book()
        book.submit_limit("a", B, 99, 5, 0.0)
        book.submit_limit("a", S, 102, 5, 0.0)
        assert book.mid() == 100.5
        assert book.spread() == 3

    def test_depth_aggregates_by_level_best_first(self):
        book = make_book()
        book.submit_limit("a", B, 100, 5, 0.0)
        book.submit_limit("b", B, 100, 3, 0.0)
        book.submit_limit("c", B, 98, 7, 0.0)
        assert book.depth(B) == [(100, 8), (98, 7)]

    def test_non_crossing_limit_generates_no_fills(self):
        book = make_book()
        book.submit_limit("a", S, 101, 5, 0.0)
        _, events = book.submit_limit("b", B, 100, 5, 0.0)
        assert fills(events) == []
        assert book.best_bid() == 100 and book.best_ask() == 101


class TestPriority:
    def test_price_priority_better_priced_maker_fills_first(self):
        book = make_book()
        book.submit_limit("cheap", S, 101, 5, 0.0)
        book.submit_limit("expensive", S, 103, 5, 0.0)
        _, events = book.submit_market("t", B, 5, 1.0)
        (fill,) = fills(events)
        assert fill.maker_owner == "cheap"
        assert fill.price == 101

    def test_time_priority_fifo_within_level(self):
        book = make_book()
        first, _ = book.submit_limit("first", S, 101, 5, 0.0)
        second, _ = book.submit_limit("second", S, 101, 5, 0.5)
        _, events = book.submit_market("t", B, 7, 1.0)
        assert [(f.maker_order_id, f.qty) for f in fills(events)] == [
            (first, 5), (second, 2)]
        # The partially-filled second order keeps its place with the remainder.
        assert book.get_order(second).qty == 3

    def test_cancel_and_replace_loses_time_priority(self):
        book = make_book()
        first, _ = book.submit_limit("first", S, 101, 5, 0.0)
        book.submit_limit("second", S, 101, 5, 0.5)
        book.cancel(first, 0.6)
        book.submit_limit("first", S, 101, 5, 0.7)  # re-enter at same price
        _, events = book.submit_market("t", B, 5, 1.0)
        assert fills(events)[0].maker_owner == "second"


class TestMatching:
    def test_execution_at_maker_price_with_price_improvement(self):
        book = make_book()
        book.submit_limit("m", S, 101, 5, 0.0)
        # Taker willing to pay 105 still executes at 101.
        _, events = book.submit_limit("t", B, 105, 5, 1.0)
        (fill,) = fills(events)
        assert fill.price == 101
        assert not isinstance(events[-1], Accepted)  # fully filled: nothing rests

    def test_taker_walks_multiple_levels(self):
        book = make_book()
        book.submit_limit("m1", S, 101, 4, 0.0)
        book.submit_limit("m2", S, 102, 4, 0.0)
        book.submit_limit("m3", S, 103, 4, 0.0)
        _, events = book.submit_market("t", B, 10, 1.0)
        assert [(f.price, f.qty) for f in fills(events)] == [(101, 4), (102, 4), (103, 2)]
        assert book.best_ask() == 103
        assert book.get_order(fills(events)[-1].maker_order_id).qty == 2

    def test_marketable_limit_fills_within_limit_then_rests(self):
        book = make_book()
        book.submit_limit("m1", S, 101, 4, 0.0)
        book.submit_limit("m2", S, 103, 4, 0.0)
        oid, events = book.submit_limit("t", B, 102, 10, 1.0)
        # Fills the 101 level, cannot touch 103, rests 6 at its own limit 102.
        assert [(f.price, f.qty) for f in fills(events)] == [(101, 4)]
        assert events[-1] == Accepted(time=1.0, order_id=oid, owner="t", side=B,
                                      price=102, qty=6)
        assert book.best_bid() == 102 and book.best_ask() == 103
        book.check_invariants()

    def test_market_order_remainder_is_cancelled_not_rested(self):
        book = make_book()
        book.submit_limit("m", S, 101, 4, 0.0)
        oid, events = book.submit_market("t", B, 10, 1.0)
        assert [(f.price, f.qty) for f in fills(events)] == [(101, 4)]
        assert events[-1] == Cancelled(time=1.0, order_id=oid, owner="t", side=B,
                                       price=None, qty=6, reason="unfilled_market")
        assert book.best_ask() is None
        assert book.best_bid() is None

    def test_market_order_on_empty_book_is_fully_cancelled(self):
        book = make_book()
        oid, events = book.submit_market("t", S, 5, 1.0)
        assert events == [Cancelled(time=1.0, order_id=oid, owner="t", side=S,
                                    price=None, qty=5, reason="unfilled_market")]

    def test_sell_side_symmetry(self):
        book = make_book()
        book.submit_limit("m1", B, 100, 4, 0.0)
        book.submit_limit("m2", B, 99, 4, 0.0)
        _, events = book.submit_limit("t", S, 99, 6, 1.0)
        assert [(f.price, f.qty) for f in fills(events)] == [(100, 4), (99, 2)]

    def test_pre_trade_mid_is_recorded_before_impact(self):
        book = make_book()
        book.submit_limit("a", B, 99, 5, 0.0)
        book.submit_limit("a", S, 101, 5, 0.0)
        _, events = book.submit_market("t", B, 5, 1.0)
        (fill,) = fills(events)
        assert fill.pre_trade_mid == 100.0  # mid before the ask was consumed


class TestCancel:
    def test_cancel_removes_order(self):
        book = make_book()
        oid, _ = book.submit_limit("a", B, 100, 5, 0.0)
        event = book.cancel(oid, 1.0)
        assert event == Cancelled(time=1.0, order_id=oid, owner="a", side=B,
                                  price=100, qty=5, reason="user")
        assert book.best_bid() is None
        assert book.get_order(oid) is None
        book.check_invariants()

    def test_cancel_unknown_or_repeated_is_noop(self):
        book = make_book()
        oid, _ = book.submit_limit("a", B, 100, 5, 0.0)
        assert book.cancel(999, 1.0) is None
        book.cancel(oid, 1.0)
        assert book.cancel(oid, 2.0) is None

    def test_cancel_after_full_fill_is_noop(self):
        book = make_book()
        oid, _ = book.submit_limit("a", S, 101, 5, 0.0)
        book.submit_market("t", B, 5, 1.0)
        assert book.cancel(oid, 2.0) is None

    def test_cancel_partially_filled_order_returns_remainder(self):
        book = make_book()
        oid, _ = book.submit_limit("a", S, 101, 5, 0.0)
        book.submit_market("t", B, 2, 1.0)
        event = book.cancel(oid, 2.0)
        assert event.qty == 3

    def test_cancel_middle_of_queue_preserves_fifo(self):
        book = make_book()
        a, _ = book.submit_limit("a", S, 101, 1, 0.0)
        b, _ = book.submit_limit("b", S, 101, 1, 0.1)
        c, _ = book.submit_limit("c", S, 101, 1, 0.2)
        book.cancel(b, 0.5)
        _, events = book.submit_market("t", B, 2, 1.0)
        assert [f.maker_order_id for f in fills(events)] == [a, c]
        book.check_invariants()


class TestValidationAndInvariants:
    @pytest.mark.parametrize("qty", [0, -1, 2.5])
    def test_bad_qty_rejected(self, qty):
        book = make_book()
        with pytest.raises(ValueError):
            book.submit_limit("a", B, 100, qty, 0.0)
        with pytest.raises(ValueError):
            book.submit_market("a", B, qty, 0.0)

    def test_non_integer_price_rejected(self):
        book = make_book()
        with pytest.raises(ValueError):
            book.submit_limit("a", B, 100.5, 5, 0.0)

    def test_level_recreated_after_emptying(self):
        """Exercises the lazy heap: a level is emptied, then re-created."""
        book = make_book()
        book.submit_limit("a", S, 101, 5, 0.0)
        book.submit_market("t", B, 5, 1.0)
        assert book.best_ask() is None
        book.submit_limit("a", S, 101, 3, 2.0)
        assert book.best_ask() == 101
        book.check_invariants()

    def test_quantity_conservation(self):
        """Total submitted = filled (counted once per side) + resting + cancelled."""
        book = make_book()
        submitted = 0
        filled = 0
        cancelled = 0
        all_events = []
        ops = [
            ("limit", "a", S, 101, 10), ("limit", "b", S, 102, 8),
            ("limit", "c", B, 100, 6), ("market", "t", B, 12),
            ("limit", "d", B, 101, 9), ("market", "t", S, 30),
        ]
        for op in ops:
            if op[0] == "limit":
                _, events = book.submit_limit(op[1], op[2], op[3], op[4], 0.0)
                submitted += op[4]
            else:
                _, events = book.submit_market(op[1], op[2], op[3], 0.0)
                submitted += op[3]
            all_events.extend(events)
            book.check_invariants()
        for e in all_events:
            if isinstance(e, Fill):
                filled += 2 * e.qty  # consumes taker and maker quantity
            elif isinstance(e, Cancelled):
                cancelled += e.qty
        resting = sum(q for side in Side for _, q in book.depth(side))
        assert submitted == filled + cancelled + resting
